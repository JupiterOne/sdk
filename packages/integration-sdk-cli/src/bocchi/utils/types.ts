import { RelationshipClass } from '@jupiterone/integration-sdk-core';

export const StepType = {
  SINGLETON: 'singleton',
  CHILD_SINGLETON: 'child-singleton',
  FETCH_ENTITIES: 'fetch-entities',
  FETCH_CHILD_ENTITIES: 'fetch-child-entities',
};

export interface Template {
  instanceConfigFields: {
    [propertyName: string]: IntegrationConfigProperty;
  };
  authentication: EndpointAuthentication | ConfigFieldAuthentication;
  baseUrl: string;
  tokenBucket?: {
    maximumCapacity: number;
    refillRate: number;
  };
  steps: Step[];
}

interface IntegrationConfigProperty {
  type?: 'string' | 'json' | 'boolean';
  mask?: Boolean;
  optional?: Boolean;
}

interface EndpointAuthentication {
  strategy: 'endpoint';
  params: {
    path: string;
    method: 'GET' | 'POST';
    body?: Record<string, any>;
    headers?: Record<string, string>;
  };
  authHeaders: AuthenticationHeaders;
}

/**
 * Authentication headers to be used in API requests.
 * The `Authorization` header is required, and is typically
 * a bearer token or API key.
 */
interface AuthenticationHeaders {
  Authorization: string;
  [headerName: string]: string;
}

interface ConfigFieldAuthentication {
  strategy: 'configField';
  authHeaders: {
    [headerName: string]: string;
  };
}

export interface Step {
  id: string;
  name: string;
  entity: {
    name: string;
    _type: string;
    _class: string | Array<string>;
    _keyPath: string;
    fieldMappings?: {
      [entityProperty: string]: string;
    };
    staticFields?: {
      [entityProperty: string]: string;
    };
  };
  parentAssociation?: {
    parentEntityType: string;
    relationshipClass: string;
  };
  request: {
    urlTemplate: string;
    method?: 'GET' | 'POST';
    params?: Record<string, any>;
  };
  response: {
    dataPath: string;
    responseType: 'SINGLETON' | 'LIST';
    nextTokenPath?: string;
  };
  directRelationships?: {
    targetKey: string;
    targetType: string;
    _class: RelationshipClass;
    direction: 'FORWARD' | 'REVERSE';
  }[];
  mappedRelationships?: {
    _class: RelationshipClass;
    direction: 'FORWARD' | 'REVERSE';
    fieldMappings?:
      | {
          sourceProperty: string;
          targetProperty: string;
        }
      | {
          targetValue: string;
          targetProperty: string;
        };
  };
  dependsOn?: string[];
}
