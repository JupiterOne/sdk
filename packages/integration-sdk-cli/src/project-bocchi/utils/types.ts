import { RelationshipClass } from '@jupiterone/integration-sdk-core';

export type StepType =
  | 'singleton'
  | 'fetch-entities'
  | 'fetch-child-entities'
  | 'fetch-relationships';

export interface Template {
  instanceConfigFields: {
    [propertyName: string]: IntegrationConfigProperty;
  };
  authentication: EndpointAuthentication | ConfigFieldAuthentication;
  baseUrl: string;
  tokenBucket?: {
    maximumCapacity: number;
    refillRate: number;
  };
  // Unimplemented
  // paginationTokenStrategy:
  steps: Step[];
}

interface IntegrationConfigProperty {
  type?: 'string' | 'json' | 'boolean';
  mask?: Boolean;
  optional?: Boolean;
}

interface EndpointAuthentication {
  strategy: 'endpoint';
  params: {
    path: string;
    method: 'GET' | 'POST';
    body?: Record<string, any>;
    headers?: Record<string, string>;
  };
  // TODO: outputHeaders
}

interface ConfigFieldAuthentication {
  strategy: 'configField';
  // TODO: outputHeaders
}

export interface Step {
  id: string;
  name: string;
  entity: {
    _type: string;
    _class: string | Array<string>;
    _keyFormat: string;
    fieldMappings?: {
      [entityProperty: string]: string;
    };
    staticFields?: {
      [entityProperty: string]: string;
    };
  };
  parentAssociation?: {
    parentEntityType: string;
    relationshipClass: string;
  };
  // TODO: api client needs to be able to do GETs and POSTs based on this
  request: {
    urlTemplate: string;
    method?: 'GET' | 'POST';
    parms?: Record<string, any>;
    tokens?: number;
  };
  response: {
    dataPath: string;
    nextTokenPath?: string;
    responseType: 'SINGLETON' | 'LIST';
  };
  directRelationships?: {
    targetKey: string;
    targetType: string;
    _class: RelationshipClass;
    direction: 'FORWARD' | 'REVERSE';
  }[];
  mappedRelationships?: {
    _class: RelationshipClass;
    direction: 'FORWARD' | 'REVERSE';
    fieldMappings:
      | {
          sourceProperty: string;
          targetProperty: string;
        }
      | {
          targetValue: string;
          targetProperty: string;
        };
  };
  dependsOn?: string[];
}
