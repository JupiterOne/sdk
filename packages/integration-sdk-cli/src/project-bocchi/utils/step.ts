import { Step, StepType } from './types';

export function determineTypeOfStep(step: Step): string {
  if (step.parentAssociation) return StepType.FETCH_CHILD_ENTITIES;
  if (!step.parentAssociation) {
    if (step.response?.responseType === 'SINGLETON') return StepType.SINGLETON;
    return StepType.FETCH_ENTITIES;
  }
  return StepType.BUILD_RELATIONSHIPS;
}
