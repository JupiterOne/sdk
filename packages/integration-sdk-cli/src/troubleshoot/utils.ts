import {
  createIntegrationInstanceForLocalExecution,
  prepareLocalStepCollection,
} from '@jupiterone/integration-sdk-runtime';
import { exec } from 'child_process';
import * as log from '../log';
import path from 'path';
import { loadConfig } from '../config';
import urlExists from 'url-exists';
import * as fs from 'fs';

export async function executeWithLogging(command: string) {
  return new Promise<void>((resolve, reject) => {
    const childProcess = exec(command);

    if (childProcess.stdout) {
      childProcess.stdout.pipe(process.stdout);
    }

    childProcess.on('exit', () => {
      resolve();
    });

    childProcess.on('error', (err) => {
      reject(err);
    });
  });
}

export async function executeValidateAuthentication(directory: string) {
  log.info(`Checking authentication using values in .env file`);
  try {
    const config = prepareLocalStepCollection(
      await loadConfig(path.join(directory, 'src')),
    );

    await import(`${directory}/src/index`).then(
      async ({ invocationConfig }) => {
        await invocationConfig.validateInvocation({
          instance: createIntegrationInstanceForLocalExecution(config),
        });
      },
    );
  } catch (err) {
    log.info(`Call to validateInvocation failed.`);
    log.info(JSON.stringify(err));
    if (err.endpoint) {
      endpointCheck(err.endpoint);
    }
  }
}

export function endpointCheck(url: string) {
  log.info(`Checking availability of endpoint ${url}`);
  try {
    urlExists(url, function (err, exists) {
      if (exists) {
        log.info(`Endpoint can be successfully reached.`);
      } else {
        log.info(
          `Checking availability of endpoint with certificate verification disabled.`,
        );
        process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
        urlExists(url, function (err, exists) {
          if (exists) {
            log.info(
              `Endpoint can be successfully reached only after disabling TLS certificate validation.`,
            );
          } else {
            log.info(`Endpoint is unreachable`);
          }
        });
      }
    });
  } catch (err) {
    log.info(err);
  }
}

export function getVersions(path: string) {
  log.info(`Checking versions from ${path}/package.json`);

  try {
    const packages = JSON.parse(
      fs.readFileSync(`${path}/package.json`, 'utf8'),
    );

    log.info(`Main package name:  ${packages.name}`);
    log.info(`Main package version:  ${packages.version}`);
    if (packages.dependencies) {
      log.info(`Dependencies:  ${JSON.stringify(packages.dependencies)}`);
    }
    if (packages.peerDependencies) {
      log.info(
        `Peer Dependencies:  ${JSON.stringify(packages.peerDependencies)}`,
      );
    }
    if (packages.devDependencies) {
      log.info(
        `Development Dependencies:  ${JSON.stringify(
          packages.devDependencies,
        )}`,
      );
    }
  } catch (err) {
    log.info(
      `An error occurred when trying to read package information from ${path}/package.json`,
    );
  }
}
