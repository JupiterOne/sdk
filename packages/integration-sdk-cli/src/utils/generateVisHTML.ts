import { defaultsDeep } from 'lodash';
import type { Node, Edge, Options } from 'vis';

export const nothingToDisplayMessage = 'There was no data found to visualize.';

const defaultOptions: Options = {
  edges: { arrows: { to: { enabled: true } } },
  physics: { barnesHut: { springLength: 300, centralGravity: 0.03 } },
};

/**
 * Creates the html to display the vis graph
 */
export function generateVisHTML(
  graphDataPath: string,
  nodeDataSets: Node[],
  edgeDataSets: Edge[],
  options?: Options,
) {
  options = defaultsDeep(options, defaultOptions);
  const displayVisualization =
    nodeDataSets.length > 0 || edgeDataSets.length > 0;

  let content = `<h1 style="text-align: center;">${graphDataPath}: ${nothingToDisplayMessage}</h1>`;

  if (displayVisualization) {
    content = `<div id="integration-graph" style="height:100vh;width:100%"></div>
      <script type="text/javascript" src="https://unpkg.com/vis-network@7.6.1/standalone/umd/vis-network.min.js"></script>
      <script type="text/javascript">
        var nodes = new vis.DataSet(${JSON.stringify(nodeDataSets)});
        var edges = new vis.DataSet(${JSON.stringify(edgeDataSets)});
        // create a network
        var container = document.getElementById("integration-graph");
        var data = {
          nodes: nodes,
          edges: edges
        };
        var options = ${JSON.stringify(options)};
        var network = new vis.Network(container, data, options);
      </script>`;
  }

  return `<!DOCTYPE html>
    <html lang="en">
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Integration SDK Graph</title>
    </head>
    <body style="height:100vh;width:100%">
      ${content}
    </body>
    </html>`;
}
