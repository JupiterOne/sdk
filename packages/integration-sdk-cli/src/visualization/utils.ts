import { v4 as uuid } from 'uuid';
import { Entity } from '@jupiterone/integration-sdk-core';

export type NodeEntity = Partial<Entity> & { nodeId: string };

/**
 * The nodeId should map back to _key, which ought to be globally unique.
 * Entities set their node ID once, and return if available.
 *
 * If nodeId is not set, first check that the _key isn't already a nodeId for another entity.
 * Duplicates will break vis.js, so we will set nodeId to a UUID (this makes isNodeDuplicate(nodeEntity) === true)
 *
 * If node is not a duplicate, return either the entity _key, or, if unavailable, return a new UUID.
 */
export function getNodeIdFromEntity(
  entity: Partial<NodeEntity>,
  existingEntities: NodeEntity[],
): string {
  if (entity.nodeId !== undefined) {
    return entity.nodeId;
  }

  if (entity._key !== undefined) {
    if (existingEntities.map((e) => e.nodeId).includes(entity._key)) {
      // duplicate key - generate UUID & continue
      // will cause isNodeIdDuplicate to return true
      return uuid();
    }
    return entity._key;
  }
  return uuid();
}

export function isNodeIdDuplicate(entity: NodeEntity): boolean {
  return entity._key !== undefined && entity._key !== entity.nodeId;
}
