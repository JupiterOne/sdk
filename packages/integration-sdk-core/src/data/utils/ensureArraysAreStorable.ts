import { AdditionalEntityProperties, AdditionalRelationshipProperties } from "..";

export function ensureNoArraysAreTooLargeToStore(
    properties: AdditionalEntityProperties | AdditionalRelationshipProperties | undefined
  ): void | never {
    for (const [key, value] of Object.entries(properties ?? {})) {
        if(Array.isArray(value)) {
            if(value.length >= 1_500) {
                throw new Error(`Property ${key} is has too many array elements to be stored: ${value.length}`)
            } else if (value.length >= 1_000) {
                console.warn({key, length: value.length} ,`Property is close to having too many array elements to be stored`)
            }
        }
    }
  }