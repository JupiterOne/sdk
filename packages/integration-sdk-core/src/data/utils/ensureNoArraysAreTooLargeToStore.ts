import { AdditionalEntityProperties } from '..';

export function ensureNoArraysAreTooLargeToStore(
  properties: AdditionalEntityProperties | undefined,
): void | never {
  for (const [key, value] of Object.entries(properties ?? {})) {
    if (Array.isArray(value)) {
      if (value.length >= 1_500) {
        throw new Error(
          `Property ${key} has too many array elements to be stored: ${value.length}`,
        );
      } else if (value.length >= 1_000) {
        console.warn(
          { propertyName: key, length: value.length },
          `Property is close to having too many array elements to be stored.`,
        );
      }
    }
  }
}
