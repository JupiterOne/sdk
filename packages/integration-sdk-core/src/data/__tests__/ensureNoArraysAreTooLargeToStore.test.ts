import { ensureNoArraysAreTooLargeToStore } from "../utils/ensureNoArraysAreTooLargeToStore"

describe('ensureNoArraysAreTooLargeToStore', () => {
    let warnSpy
    beforeEach(() => {
        warnSpy = jest.spyOn(global.console, 'warn')
    })
    afterEach(() => {
        warnSpy.mockRestore()
    });

    it('should error out if there is a property that is an array with more than 1,500 elements', () => {
        const properties = {
            array: new Array(1500).fill(Math.random())
        }
        expect(() => ensureNoArraysAreTooLargeToStore(properties)).toThrowErrorMatchingSnapshot()
    })
    it('should warn if there is a property that is an array with more than 1,000 elements, but less than 1,500 elements', () => {
        let properties = {
            array: new Array(1000).fill(Math.random())
        }
        ensureNoArraysAreTooLargeToStore(properties)
        expect(warnSpy).toHaveBeenCalled()

        warnSpy.mockReset()
        properties = {
            array: new Array(1499).fill(Math.random())
        }
        ensureNoArraysAreTooLargeToStore(properties)
        expect(warnSpy).toHaveBeenCalled()
    })
    it('should do nothing if an array has fewer than 1,000 elements', () => {
        const properties = {
            array: new Array(999).fill(Math.random())
        }
        ensureNoArraysAreTooLargeToStore(properties)
        expect(warnSpy).not.toHaveBeenCalled()
    })
    it('should do nothing for properties that are not arrays', () => {
        const warnSpy = jest.spyOn(global.console, 'warn')
        const properties = {
            string: 'string',
            number: 1,
            undefined: undefined,
            null: null
        }
        ensureNoArraysAreTooLargeToStore(properties)
        expect(warnSpy).not.toHaveBeenCalled()
    })
})
