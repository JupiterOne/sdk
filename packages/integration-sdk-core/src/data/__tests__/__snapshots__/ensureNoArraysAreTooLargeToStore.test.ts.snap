// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ensureNoArraysAreTooLargeToStore should error out if there is a property that is an array with more than 1,500 elements 1`] = `"Property array has too many array elements to be stored: 1500"`;
